File: HoYa/Entities/WorkFlow.cs
using Newtonsoft.Json;
using System;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace HoYa.Entities
{
    public class WorkFlow : Definition
    {
        public virtual Guid? FirstId { get; set; }
        [ForeignKey("FirstId")]
        [JsonIgnore]
        public virtual Step First { get; set; }
        public string Code { get; set; }
        public string Value { get; set; }
        public virtual Guid? StatusId { get; set; }
        [ForeignKey("StatusId")]
        public virtual Option Status { get; set; }
        public virtual Guid? FinalId { get; set; }
        [ForeignKey("FinalId")]
        [JsonIgnore]
        public virtual Step Final { get; set; }
    }

    public class Step : Base
    {
        public virtual Guid? OwnerId { get; set; }
        [ForeignKey("OwnerId")]
        public virtual WorkFlow Owner { get; set; }
        public virtual Guid? TypeId { get; set; }
        [ForeignKey("TypeId")]
        public virtual Option Type { get; set; }
        public decimal? Sort { get; set; }
        public string Value { get; set; }
        public string View { get; set; }
        public decimal? SubmitRatio { get; set; }
        public string Component { get; set; }
    }

    public class Redirect : Relation<Step, Step>
    {
        public decimal? TrueRatio { get; set; }//(=100)=AND (>0)=OR
    }

    public class RedirectCondition : Relation<Redirect, Condition>
    {
        public decimal? Sort { get; set; }
    }

    public class Condition : Definition
    {
    }

    public class StepGroup : Relation<Step, Inventory>
    {
        public float? Sort { get; set; }
    }

    public class StepRelationship : Relation<Step, Option>//這裡的Option是Of的關係 例如主管 或 代理人
    {
        public decimal? Sort { get; set; }
        public virtual Guid? OfId { get; set; }
        [ForeignKey("OfId")]
        public virtual Step Of { get; set; }//開單者 或 前一關卡者
    }

    public class Activity : Relation<Process, Step>
    {
        public virtual Guid? PreviousId { get; set; }
        [ForeignKey("PreviousId")]
        public virtual Activity Previous { get; set; }
        public virtual Guid? NextId { get; set; }
        [ForeignKey("NextId")]
        [JsonIgnore]
        public virtual Activity Next { get; set; }
        public virtual Guid? ArchivedParticipateId { get; set; }
        [ForeignKey("ArchivedParticipateId")]
        [JsonIgnore]
        public virtual Participate ArchivedParticipate { get; set; }
    }

    public class Participate :Base
    {
        public virtual Guid? OwnerId { get; set; }
        [ForeignKey("OwnerId")]
        [JsonIgnore]
        public virtual Activity Owner { get; set; }
        public DateTime? StartDate { get; set; }

        public DateTime? EndDate { get; set; }

        public DateTime? ArchivedDate { get; set; }

        public virtual Guid? ArchivedById { get; set; }
        [ForeignKey("ArchivedById")]
        [JsonIgnore]
        public virtual Inventory ArchivedBy { get; set; }

        public virtual Guid? ParticipantId { get; set; }
        [ForeignKey("ParticipantId")]
        public virtual Inventory Participant { get; set; }

        public virtual Guid? RedirectId { get; set; }
        [ForeignKey("RedirectId")]
        public virtual Redirect Redirect { get; set; }

        public DateTime? ReadedDate { get; set; }
        public Participate()
        {
            StartDate = DateTime.Now;
        }
    }

    public class Judgment :Base
    {
        public virtual Guid? OwnerId { get; set; }
        [ForeignKey("OwnerId")]
        [JsonIgnore]
        public virtual Participate Owner { get; set; }
        public virtual Guid? ConditionId { get; set; }
        [ForeignKey("ConditionId")]
        public virtual Condition Condition { get; set; }
    }

    public class Process : Instance
    {
        public virtual Guid? WorkFlowId { get; set; }
        [ForeignKey("WorkFlowId")]
        public virtual WorkFlow WorkFlow { get; set; }
        public virtual Guid? ParentId { get; set; }
        [ForeignKey("ParentId")]
        public virtual Process Parent { get; set; }
        public int? ReEdit { get; set; }
        public virtual Guid? PreviousId { get; set; }
        [ForeignKey("PreviousId")]
        public virtual Activity Previous { get; set; }
        public virtual Guid? CurrentId { get; set; }
        [ForeignKey("CurrentId")]
        public virtual Activity Current { get; set; }
    }
}

File: HoYa.Web/src/@entities/workflow.ts
import { Definition, Relation, Detail, Option, Instance } from "./entity";
import { Group } from "./group";
import { Inventory } from "./inventory";


export class WorkFlow extends Definition {
    firstId: string;
    first: Step;
    finalId: string;
    final: Step;
    constructor(id?: string, firstId?: string) {
        super();
        this.id = id;
        this.firstId = firstId;
    }
}

export class Step extends Detail<WorkFlow>
{
    view: string;
    component: string;
    constructor() {
        super();
    }
}
export class StepGroup extends Relation<Step, Group>
{
    constructor() {
        super();
    }
}

export class Redirect extends Relation<Step, Step>
{
    trueRatio: number;//(=100)=AND (>0)=OR
    _errorirectConditions: RedirectCondition[];
    constructor() {
        super();
    }
}

export class RedirectCondition extends Relation<Redirect, Condition>
{
    sort: number;
    constructor() {
        super();
    }
}


export class StepRelationship extends Relation<Step, Option>//這裡的Option是Of的關係 例如主管 或 代理人
{
    ofId: string;
    of: Option;//開單者 或 前一關卡者
    constructor() {
        super();
    }
}


export class Activity extends Relation<Process, Step>
{
    archivedParticipate: Participate;
    archivedParticipateId: string;
    previousId: string;
    previous: Activity;
    nextId: string;
    next: Activity;
    componentId: string;
    _ignore: boolean;
    constructor() {
        super();
    }
}

export class Condition extends Definition {
    constructor() {
        super();

    }
}

export class Participate extends Detail<Activity>
{
    startDate: Date;
    endDate: Date;
    archivedDate: Date;
    archivedById: string;
    archivedBy: Inventory;
    participantId: string;
    participant: Inventory;
    redirectId: string;
    redirect: Redirect;
    actionId: string;
    action: Option;
    constructor(ownerId?: string, participantId?: string) {
        super();
        this.startDate = new Date();
        this.ownerId = ownerId;
        this.participantId = participantId;
    }
}

export class Judgment extends Detail<Participate>
{
    conditionId: string;
    condition: Condition;
    constructor() {
        super();
    }
}

export class Process extends Instance {
    previousId: string;
    previous: Activity;
    currentId: string;
    current: Activity;
    workFlowId: string;
    workFlow: WorkFlow;
    parentId;
    parent: Process;
    _activitiesCount: number;
    reEdit: number;
    constructor(workFlow?: WorkFlow) {
        super();
        this.reEdit = 1;
        this.workFlow = workFlow;
        this.workFlowId = workFlow.id;
    }
}